File: DeckBuilderElements/handBuilders.js
import {Ally} from '../models/classes.js'
import {config, print} from '../config.js'

export const HAND_BUILDER_TYPES = {
    RANDOM: 'RandomHandBuilder',
    N_OF_EACH: 'NOfEachHandBuilder'
}

export class BaseHandBuilder {
    static makeHandBuilder = function(handBuilderType){
        switch (handBuilderType) {
            case HAND_BUILDER_TYPES.RANDOM:
                return new RandomHandBuilder(config.MIN_HERO_ALLY_ATTACK, config.MAX_HERO_ALLY_ATTACK)
            case HAND_BUILDER_TYPES.N_OF_EACH:
                return new NOfEachHandBuilder(config.MIN_HERO_ALLY_ATTACK, config.MAX_HERO_ALLY_ATTACK)
            default:
                throw new Error('Invalid hand builder type')
        }
    }
}

export class RandomHandBuilder extends BaseHandBuilder{
    constructor(minDamage, maxDamage) {
        super();
        this.minDamage = minDamage
        this.maxDamage = maxDamage
    }
    
    buildHand = function(deck, handSize) {
        var hand = []
        for (var i = 0; i < handSize; i++) {
            var index = Math.floor(Math.random() * deck.length)
            hand.push(deck[index])
            deck.splice(index, 1)
        }
        return hand
    }
}

export class NOfEachHandBuilder extends BaseHandBuilder{
    constructor(minDamage, maxDamage) {
        super();
        this.minDamage = minDamage
        this.maxDamage = maxDamage
    }
    
    buildHand = function(deck, handSize){
        var hand = []
        var n = Math.floor(handSize / (this.maxDamage - this.minDamage + 1))
        for (var i = this.minDamage; i <= this.maxDamage; i++) {
            var card = deck.find(n => n.attack == i)
            if (!card) {
                card = new Ally('Ally Damage ' + i + ' Cost: ' + cost, '', 0, i, 0)
            }
            for (var j = 0; j < n; j++) {
                hand.push(card)
            }
        }
        while(hand.length < handSize){
            hand.push(new Ally('Ally Damage ' + 0 + ' Cost: ' + cost, '', 0, 0, 0))
        }
        return hand
    }
}

export class DrawFromTopOfDeckHandBuilder extends BaseHandBuilder{
    constructor(minDamage, maxDamage) {
        super();
        this.minDamage = minDamage
        this.maxDamage = maxDamage
    }
    
    buildHand = function(deck, handSize){
        var hand = []
        while(hand.length < handSize){
            var card = deck.shift()
            if (card.attack >= this.minDamage && card.attack <= this.maxDamage) {
                hand.push(card)
            }else{
                hand.push(new Ally('Ally Damage ' + i + ' Cost: ' + cost, '', 0, i, 0))
                deck.push(card)
            }
        }
        return hand
    }
}

File: python/DeckBuilderElements/handBuilders.py
import tensorflow as tf
from models.killable import Killable
from config import config, log
import random

class HAND_BUILDER_TYPES:
    RANDOM= 'RandomHandBuilder'
    N_OF_EACH= 'NOfEachHandBuilder'

class BaseHandBuilder:
    def makeHandBuilder(handBuilderType):
        if handBuilderType == HAND_BUILDER_TYPES.RANDOM:
            return RandomHandBuilder(config.MIN_HERO_ALLY_ATTACK, config.MAX_HERO_ALLY_ATTACK)
        elif handBuilderType == HAND_BUILDER_TYPES.N_OF_EACH:
            return NOfEachHandBuilder(config.MIN_HERO_ALLY_ATTACK, config.MAX_HERO_ALLY_ATTACK)
        else:
            raise Exception('Invalid hand builder type')

class RandomHandBuilder(BaseHandBuilder):
    def __init__(self, minDamage, maxDamage):
        self.minDamage = minDamage
        self.maxDamage = maxDamage
    
    def buildHand(self, deck, handSize):
        hand = []
        for i in range(handSize):
            index = random.randint(0, len(deck) - 1)
            hand.append(deck[index])
            deck.pop(index)
        return hand

class NOfEachHandBuilder(BaseHandBuilder):
    def __init__(self, minDamage, maxDamage):
        self.minDamage = minDamage
        self.maxDamage = maxDamage
    
    def buildHand(self, deck, handSize):
        hand = []
        n = handSize // (self.maxDamage - self.minDamage + 1)
        for i in range(self.minDamage, self.maxDamage + 1):
            card = next((n for n in deck if n.attack == i), None)
            if not card:
                card = Killable('Ally Damage ' + str(i) + ' Cost: ' + str(config.COST), '', 0, i, 0)
            for j in range(n):
                hand.append(card)
        while len(hand) < handSize:
            hand.append(Killable('Ally Damage ' + str(0) + ' Cost: ' + str(config.COST), '', 0, 0, 0))
        return hand

class DrawFromTopOfDeckHandBuilder(BaseHandBuilder):
    def __init__(self, minDamage, maxDamage):
        self.minDamage = minDamage
        self.maxDamage = maxDamage
    
    def buildHand(self, deck, handSize):
        hand = []
        while len(hand) < handSize:
            card = deck.pop(0)
            if card.attack >= self.minDamage and card.attack <= self.maxDamage:
                hand.append(card)
            else:
                hand.append(Killable('Ally Damage ' + str(0) + ' Cost: ' + str(config.COST), '', 0, 0, 0))
                deck.append(card)
        return hand




